<a href="https://colab.research.google.com/github/EdgaRojas/MachineLearning/blob/main/Unidad%203/Proyecto/ProyectoU3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PROYECTO UNIDAD 3
## Alumno: Edgar Manuel Rojas Gurrola


## KNN

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

# Importamos librerias necesarias para knn
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_recall_curve

In [ ]:

prestamos_df = pd.read_csv('prestamos_ok.csv')
prestamos_df.head()


Unnamed: 0  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0           0       2400         2400           2400.0  36 months     15.96   
1           1      10000        10000          10000.0  36 months     13.49   
2           2       3000         3000           3000.0  36 months     18.64   
3           3       5600         5600           5600.0  60 months     21.28   
4           4       5375         5375           5350.0  60 months     12.69   

   installment grade sub_grade            emp_title  ... disbursement_method  \
0        84.33     C        C5                  NaN  ...                Cash   
1       339.31     C        C1  AIR RESOURCES BOARD  ...                Cash   
2       109.43     E        E1      MKC Accounting   ...                Cash   
3       152.39     F        F2                  NaN  ...                Cash   
4       121.45     B        B5            Starbucks  ...                Cash   

  debt_settlement_flag  debt_settlement_flag_date issue_year repaid  \
0                    N                        NaN       2011      1   
1                    N                        NaN       2011      1   
2                    N                        NaN       2011      1   
3                    N                        NaN       2011      0   
4                    N                        NaN       2011      0   

  loan_term_year purpose_code home_ownership_code grade_code addr_state_code  
0              3           11                   4          2               2  
1              3            9                   4          2               0  
2              3            0                   4          4               0  
3              5           11                   3          5               0  
4              5            9                   4          1               5  

[5 rows x 66 columns]

In [ ]:
X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

# Variable objetivo o variable a predecir
y = prestamos_df["repaid"]

Se eligen estas variables porque despues de observar la tabla de mayor importancia de cada variable decidi utilizar las primeras

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.3, random_state=42)
print("Dimensiones de X_train:", X_train.shape)
print("Dimensiones de X_test:", X_test.shape)

Dimensiones de X_train: (13935, 27)
Dimensiones de X_test: (5973, 27)


In [ ]:
y_pred = knn1.predict(X_test)
print("Reporte de Clasificación:\n", classification_report(y_test, y_pred))
print("Matriz de Confusión:\n", confusion_matrix(y_test, y_pred))

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.18      0.16      0.17       930
           1       0.85      0.87      0.86      5043

    accuracy                           0.76      5973
   macro avg       0.52      0.51      0.52      5973
weighted avg       0.75      0.76      0.75      5973

Matriz de Confusión:
 [[ 150  780]
 [ 666 4377]]


In [ ]:
import collections
from imblearn.over_sampling import SMOTE

print('Distribución de clases antes del oversampling (y_train): %s' % collections.Counter(y_train))

sm = SMOTE(random_state=42)
X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

print('Distribución de clases después del oversampling (y_train_res): %s' % collections.Counter(y_train_res))

Distribución de clases antes del oversampling (y_train): Counter({1: 11922, 0: 2013})
Distribución de clases después del oversampling (y_train_res): Counter({1: 11922, 0: 11922})


In [ ]:
# Reentrenar el modelo KNN con los datos sobremuestreados
knn_resampled = KNeighborsClassifier(n_neighbors=1)
knn_resampled.fit(X_train_res, y_train_res)

print("Precisión del clasificador en fase de entrenamiento (después de oversampling):", knn_resampled.score(X_train_res, y_train_res))

Precisión del clasificador en fase de entrenamiento (después de oversampling): 1.0


In [ ]:
# Evaluar el modelo reentrenado en el conjunto de prueba original
y_pred_resampled = knn_resampled.predict(X_test)

print("Reporte de Clasificación (después de oversampling):\n", classification_report(y_test, y_pred_resampled, target_names=["No Pagado", "Pagado"]))
print("Matriz de Confusión (después de oversampling):\n", confusion_matrix(y_test, y_pred_resampled))

Reporte de Clasificación (después de oversampling):
               precision    recall  f1-score   support

   No Pagado       0.16      0.24      0.20       930
      Pagado       0.85      0.78      0.81      5043

    accuracy                           0.69      5973
   macro avg       0.51      0.51      0.50      5973
weighted avg       0.74      0.69      0.71      5973

Matriz de Confusión (después de oversampling):
 [[ 224  706]
 [1134 3909]]


In [ ]:
knn1 = KNeighborsClassifier(n_neighbors = 1)
knn1.fit(X_train, y_train)
print("Precision del clasificador en fase de entrenamiento:", knn1.score(X_train, y_train))

Precision del clasificador en fase de entrenamiento: 1.0


In [ ]:

y_pred = knn1.predict(X_test)
print("\nReporte de métricas del clasificador con 1 vecino: \n",
      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))

print(f'\nMatriz Confusion con 1 vecino:\n', confusion_matrix(y_test, y_pred ))


Reporte de métricas del clasificador con 1 vecino: 
               precision    recall  f1-score   support

   No Pagado       0.18      0.16      0.17       930
      Pagado       0.85      0.87      0.86      5043

    accuracy                           0.76      5973
   macro avg       0.52      0.51      0.52      5973
weighted avg       0.75      0.76      0.75      5973


Matriz Confusion con 1 vecino:
 [[ 150  780]
 [ 666 4377]]


- Se dectectan 150 impagos reales de 930
- Se obtuvo una Precisión del 76% en la detección de pagos

In [ ]:

knn5 = KNeighborsClassifier(n_neighbors = 5)
knn5.fit(X_train, y_train)
print("Precision del clasificador en fase de entrenamiento", knn5.score(X_train, y_train) )


Precision del clasificador en fase de entrenamiento 0.8630785791173304


In [ ]:

y_pred = knn5.predict(X_test)

print("\nReporte de métricas del clasificador con 5 vecinos: \n",
      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))
print(f'\nMatriz de Confusion con 5 vecinos:\n', confusion_matrix(y_test, y_pred ))


Reporte de métricas del clasificador con 5 vecinos: 
               precision    recall  f1-score   support

   No Pagado       0.22      0.04      0.06       930
      Pagado       0.85      0.98      0.91      5043

    accuracy                           0.83      5973
   macro avg       0.53      0.51      0.48      5973
weighted avg       0.75      0.83      0.77      5973


Matriz de Confusion con 5 vecinos:
 [[  33  897]
 [ 120 4923]]


Se mejoro la precision de un 76% a un 83%
Solo detecta 33 impagos reales de 930

In [ ]:

knn10 = KNeighborsClassifier(n_neighbors = 10)
knn10.fit(X_train, y_train)

print("Precision del clasificador en fase de entrenamiento", knn10.score(X_train, y_train) )


Precision del clasificador en fase de entrenamiento 0.8552565482597776


In [ ]:

y_pred = knn10.predict(X_test)

print("\nReporte de métricas del clasificador con 10 vecinos: \n",
      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))

print(f'Matriz de Confusion con 10 vecinos:\n', confusion_matrix(y_test, y_pred ))


Reporte de métricas del clasificador con 10 vecinos: 
               precision    recall  f1-score   support

   No Pagado       0.32      0.02      0.04       930
      Pagado       0.85      0.99      0.91      5043

    accuracy                           0.84      5973
   macro avg       0.58      0.51      0.48      5973
weighted avg       0.76      0.84      0.78      5973

Matriz de Confusion con 10 vecinos:
 [[  20  910]
 [  43 5000]]


La precision aumento 1%
Solo detecta 20 impagos reales de 930

In [ ]:
from sklearn.preprocessing import StandardScaler

X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

y = prestamos_df["repaid"]

# Dividir Datos, stratify es para que mantenga la misma proporción de clases en ambos conjuntos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42, stratify=y)

# Normalizar variables (KNN es sensible a la escala)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Use StandardScaler para que el modelo KNN pudiera funcionar de manera óptima al calcular las distancias entre los vecinos.

In [ ]:
# Entrenar KNN
knn_scaled = KNeighborsClassifier(n_neighbors=3)
knn_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = knn_scaled.predict(X_test_scaled)
print("\nReporte de clasificación datos normalizados y 3 vecinos:")
print(classification_report(y_test, y_pred_scaled, target_names=["No Pagado", "Pagado"]))

print ("\nMatriz de confusión con datos normalizados y 3 vecinos:\n", confusion_matrix(y_test, y_pred_scaled), "\n")


Reporte de clasificación datos normalizados y 3 vecinos:
              precision    recall  f1-score   support

   No Pagado       0.22      0.10      0.14      1177
      Pagado       0.86      0.93      0.89      6787

    accuracy                           0.81      7964
   macro avg       0.54      0.52      0.52      7964
weighted avg       0.76      0.81      0.78      7964


Matriz de confusión con datos normalizados y 3 vecinos:
 [[ 122 1055]
 [ 445 6342]] 



- Después de la normalizacion usando 3 vecinos el porcentaje de precisión obtenido es de 81%
- 122 impagos reales de 1177

In [ ]:
# Entrenar KNN
knn_scaled = KNeighborsClassifier(n_neighbors=5)
knn_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = knn_scaled.predict(X_test_scaled)
print("\nReporte de clasificación datos normalizados y 5 vecinos:")
print(classification_report(y_test, y_pred_scaled, target_names=["No Pagado", "Pagado"]))

print ("\nMatriz de confusión con datos normalizados y 5 vecinos:\n", confusion_matrix(y_test, y_pred_scaled), "\n")


Reporte de clasificación datos normalizados y 5 vecinos:
              precision    recall  f1-score   support

   No Pagado       0.22      0.06      0.09      1177
      Pagado       0.86      0.96      0.91      6787

    accuracy                           0.83      7964
   macro avg       0.54      0.51      0.50      7964
weighted avg       0.76      0.83      0.79      7964


Matriz de confusión con datos normalizados y 5 vecinos:
 [[  69 1108]
 [ 239 6548]] 



- Usando 5 vecinos se mejoró la precisión un 2%
-Solo 69 impagos reales de 1177

In [ ]:
# Entrenar KNN
knn_scaled = KNeighborsClassifier(n_neighbors=10)
knn_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = knn_scaled.predict(X_test_scaled)
print("\nReporte de clasificación datos normalizados y 10 vecinos:")
print(classification_report(y_test, y_pred_scaled, target_names=["No Pagado", "Pagado"]))

print ("\nMatriz de confusión con datos normalizados y 10 vecinos:\n", confusion_matrix(y_test, y_pred_scaled), "\n")


Reporte de clasificación datos normalizados y 10 vecinos:
              precision    recall  f1-score   support

   No Pagado       0.29      0.04      0.08      1177
      Pagado       0.86      0.98      0.91      6787

    accuracy                           0.84      7964
   macro avg       0.57      0.51      0.50      7964
weighted avg       0.77      0.84      0.79      7964


Matriz de confusión con datos normalizados y 10 vecinos:
 [[  52 1125]
 [ 128 6659]] 



- Se mejoró la precisión 1%
- los impagos reales bajaron a 52

- El algoritmo lo que hace es comparar variables de cada dato con otros parecidos y de esa manera identificar si esa persona pagara o no.
- Ventajas: Es flexible, simple e intuitivo
- Limitaciones: Es lento con grandes cantidades de datos, es necesario normalizar datos para mejor funcionamiento
- El modelo aun no detecta
bien los no pagados
- Mejorar la eleccion de variables y normalizar los datos de esas variables

## Regresión Logística

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

# Importamos librerias necesarias para regresion logistica y evaluacion de modelos
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_recall_curve

In [ ]:
prestamos_df = pd.read_csv('prestamos_ok.csv')
prestamos_df.head()

Unnamed: 0  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0           0       2400         2400           2400.0  36 months     15.96   
1           1      10000        10000          10000.0  36 months     13.49   
2           2       3000         3000           3000.0  36 months     18.64   
3           3       5600         5600           5600.0  60 months     21.28   
4           4       5375         5375           5350.0  60 months     12.69   

   installment grade sub_grade            emp_title  ... disbursement_method  \
0        84.33     C        C5                  NaN  ...                Cash   
1       339.31     C        C1  AIR RESOURCES BOARD  ...                Cash   
2       109.43     E        E1      MKC Accounting   ...                Cash   
3       152.39     F        F2                  NaN  ...                Cash   
4       121.45     B        B5            Starbucks  ...                Cash   

  debt_settlement_flag  debt_settlement_flag_date issue_year repaid  \
0                    N                        NaN       2011      1   
1                    N                        NaN       2011      1   
2                    N                        NaN       2011      1   
3                    N                        NaN       2011      0   
4                    N                        NaN       2011      0   

  loan_term_year purpose_code home_ownership_code grade_code addr_state_code  
0              3           11                   4          2               2  
1              3            9                   4          2               0  
2              3            0                   4          4               0  
3              5           11                   3          5               0  
4              5            9                   4          1               5  

[5 rows x 66 columns]

In [ ]:
X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

# Variable objetivo o variable a predecir
y = prestamos_df["repaid"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((13935, 7), (5973, 7), (13935,), (5973,))

In [ ]:

clf2 = LogisticRegression(random_state=0, class_weight="balanced")
clf2.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(class_weight='balanced', random_state=0)

In [ ]:
print("Exactitud del modelo de entrenamiento:", clf2.score(X_train, y_train))

Exactitud del modelo de entrenamiento: 0.5289558665231432


In [ ]:
# Realizar una prediccion con los datos de prueba
y_pred2 = clf2.predict(X_test)

# Crear un informe de texto que muestre las principales métricas de clasificación.
print("\nReporte de métricas del clasificador: \n",
      classification_report(y_test, y_pred2, target_names=["No Pagado", "Pagado"]))

# Coeficiente de determinación
print( "Precisión:", clf2.score(X_test, y_test) )


Reporte de métricas del clasificador: 
               precision    recall  f1-score   support

   No Pagado       0.20      0.69      0.31       930
      Pagado       0.90      0.50      0.64      5043

    accuracy                           0.53      5973
   macro avg       0.55      0.59      0.48      5973
weighted avg       0.79      0.53      0.59      5973

Precisión: 0.5287125397622635


Usando la regresión logística bajo bastante la precisión de detección

In [ ]:
matriz_balanceo = confusion_matrix(y_test, y_pred2 )
print(f'Matriz Confusion:\n', matriz_balanceo)

Matriz Confusion:
 [[ 640  290]
 [2525 2518]]


Se obtuvo un 53% de precision
640 impagos reales de 930

In [ ]:
from sklearn.preprocessing import StandardScaler

X = prestamos_df[['funded_amnt', "int_rate", "grade_code", 'purpose_code', 'addr_state_code',
                 'home_ownership_code', 'annual_inc', 'dti', 'revol_util',
                 'pub_rec_bankruptcies']]

y = prestamos_df["repaid"]

# stratify es para que mantenga la misma proporción de clases en ambos conjuntos
# Usamos los mismos parametros de tamaño y random usados anteriormente para que sea justa la comparación
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.4, random_state=42, stratify=y)

# Normalizar variables (Regresión Logística (RL) es sensible a la escala)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entrenar modelo RL escalado
RL_scaled = LogisticRegression(random_state=0, class_weight='balanced')
RL_scaled.fit(X_train_scaled, y_train)

# Probar y evaluar el modelo RL escalado
y_pred_scaled = RL_scaled.predict(X_test_scaled)

print("\nReporte de clasificación:")
print(classification_report(y_test, y_pred_scaled, target_names=["No Pagado", "Pagado"]))


Reporte de clasificación:
              precision    recall  f1-score   support

   No Pagado       0.22      0.61      0.32      1177
      Pagado       0.90      0.62      0.73      6787

    accuracy                           0.61      7964
   macro avg       0.56      0.61      0.53      7964
weighted avg       0.80      0.61      0.67      7964



In [ ]:

print("\nMatriz de Confusión:")
print(confusion_matrix(y_test, y_pred_scaled))


Matriz de Confusión:
[[ 719  458]
 [2611 4176]]


Despues de la normalizacion el porcentaje de precision subio un 8%
719 impagos reales de 1177

- La regresion consiste en agarrar un dato y obtener una probabilidad de que pertenezca a una variable binaria como pagado o no pagado
- ventajas: usado para gran cantidad de datos y es menos propensa a un sobreajuste si nos son demasiadas variables predictoras
- Limitaciones: Asume linealidas entre las variables, requiere datos balanceados
- Baja precision en no pagado detecta 20% bien y lo demas si pagaron en realidad pero los catalogo como no pagados
- Hay un desbalance entre las variables que se catalogan como pagadas y las que son no pagadas siendo las pagadas la de mayor cantidad, y tal vez no seleccione las variables correctas o son demasiadas
- Mejorar la seleccion de variables y que los datos del dataset esten mas balanceados

## Árbol de Decisión


In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

# Importamos librerias necesarias para Árboles de decisión
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_recall_curve


In [ ]:
prestamos_df = pd.read_csv('prestamos_ok.csv')
prestamos_df.head()

Unnamed: 0  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0           0       2400         2400           2400.0  36 months     15.96   
1           1      10000        10000          10000.0  36 months     13.49   
2           2       3000         3000           3000.0  36 months     18.64   
3           3       5600         5600           5600.0  60 months     21.28   
4           4       5375         5375           5350.0  60 months     12.69   

   installment grade sub_grade            emp_title  ... disbursement_method  \
0        84.33     C        C5                  NaN  ...                Cash   
1       339.31     C        C1  AIR RESOURCES BOARD  ...                Cash   
2       109.43     E        E1      MKC Accounting   ...                Cash   
3       152.39     F        F2                  NaN  ...                Cash   
4       121.45     B        B5            Starbucks  ...                Cash   

  debt_settlement_flag  debt_settlement_flag_date issue_year repaid  \
0                    N                        NaN       2011      1   
1                    N                        NaN       2011      1   
2                    N                        NaN       2011      1   
3                    N                        NaN       2011      0   
4                    N                        NaN       2011      0   

  loan_term_year purpose_code home_ownership_code grade_code addr_state_code  
0              3           11                   4          2               2  
1              3            9                   4          2               0  
2              3            0                   4          4               0  
3              5           11                   3          5               0  
4              5            9                   4          1               5  

[5 rows x 66 columns]

In [ ]:
X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

# Variable objetivo o variable a predecir
y = prestamos_df["repaid"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size= 0.3, stratify=y )


In [ ]:
# Crear el modelo de árbol de clasificación
tree_model = DecisionTreeClassifier(criterion='gini', max_depth=5, random_state=42)

# Entrenar el modelo
tree_model.fit(X_train, y_train)

# Precisión del modelo en la fase de entrenamiento
print("Precision del clasificador en fase de entrenamiento", tree_model.score(X_train, y_train) )

Precision del clasificador en fase de entrenamiento 0.8534625044851094


Se establece los parámetros por defecto del modelo

In [ ]:

y_pred = tree_model.predict(X_test)
print("\nReporte de métricas del clasificador de Arboles de Decisión: \n",
      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))

print(f'\nMatriz Confusion de Arboles de Decisión:\n', confusion_matrix(y_test, y_pred ))


Reporte de métricas del clasificador de Arboles de Decisión: 
               precision    recall  f1-score   support

   No Pagado       0.26      0.01      0.01       883
      Pagado       0.85      1.00      0.92      5090

    accuracy                           0.85      5973
   macro avg       0.56      0.50      0.47      5973
weighted avg       0.77      0.85      0.78      5973


Matriz Confusion de Arboles de Decisión:
 [[   5  878]
 [  14 5076]]


- Con los valores por defecto se obtuvo un 85% de precision
- solo 5 impagos reales de 883

In [ ]:
# Crear modelo con balanceo de clases
tree_model2 = DecisionTreeClassifier(criterion='gini', max_depth=5, class_weight='balanced', random_state=42, )

# Entrenar el modelo
tree_model2.fit(X_train, y_train)

# Realizar fase de pruena
y_pred2 = tree_model2.predict(X_test)

# Evaluar el modelo
print("\nReporte de métricas del clasificador de Arboles de Decisión: \n",
      classification_report(y_test, y_pred2, target_names=["No Pagado", "Pagado"]))

print("\n Matriz de Confusión:")
print(confusion_matrix(y_test, y_pred2))



Reporte de métricas del clasificador de Arboles de Decisión: 
               precision    recall  f1-score   support

   No Pagado       0.20      0.72      0.31       883
      Pagado       0.91      0.51      0.65      5090

    accuracy                           0.54      5973
   macro avg       0.56      0.61      0.48      5973
weighted avg       0.81      0.54      0.60      5973


 Matriz de Confusión:
[[ 633  250]
 [2510 2580]]


Bajo demasiado el porcentaje de precision y aumentaron los impagos reales a 633 de 883

In [ ]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [5, 7, 10, None],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 5, 10],
    'class_weight': [None, 'balanced'],
    'ccp_alpha': [0.0, 0.001, 0.01]
}

grid_search = GridSearchCV(
    DecisionTreeClassifier(random_state=42),
    param_grid,
    scoring='recall_macro',   # o 'f2' si te interesa más los impagos
    cv=5,
    n_jobs=-1,
    verbose=2
)

grid_search.fit(X_train, y_train)
print("Mejores parámetros:", grid_search.best_params_)

# Evaluación del mejor modelo
best_model = grid_search.best_estimator_

y_pred = best_model.predict(X_test)

print("\nReporte de Clasificación (modelo optimizado para detectar impagos):\n")
print(classification_report(y_test, y_pred))

print("Matriz de Confusión:")
print(confusion_matrix(y_test, y_pred))

Fitting 5 folds for each of 768 candidates, totalling 3840 fits
Mejores parámetros: {'ccp_alpha': 0.001, 'class_weight': 'balanced', 'criterion': 'gini', 'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2}

Reporte de Clasificación (modelo optimizado para detectar impagos):

              precision    recall  f1-score   support

           0       0.20      0.73      0.32       883
           1       0.92      0.51      0.65      5090

    accuracy                           0.54      5973
   macro avg       0.56      0.62      0.49      5973
weighted avg       0.81      0.54      0.61      5973

Matriz de Confusión:
[[ 642  241]
 [2494 2596]]


Se obtuvieron mejores resultados con los parametros por defecto y sin balanceo

- El modelo árboles de decisión sirve dividiendo los datos en conjunto pequeños para su análisis
- Ventajas: Fácil de entender, requiere poca preparación de datos, es eficiente.
- Limitaciones: sobreajuste e inestabilidad.
- Detecta muy pocos casos de préstamos no pagados
- Desbalance de clases y sobreajuste
- Mejor selección de variables, balance entre clases de los datos del dataset
-GridSearchCV/RandomizedSearchCV

## Random Forest


In [ ]:

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

# Importamos librerias necesarias para Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_recall_curve


In [ ]:
prestamos_df = pd.read_csv('prestamos_ok.csv')
prestamos_df.head()

Unnamed: 0  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0           0       2400         2400           2400.0  36 months     15.96   
1           1      10000        10000          10000.0  36 months     13.49   
2           2       3000         3000           3000.0  36 months     18.64   
3           3       5600         5600           5600.0  60 months     21.28   
4           4       5375         5375           5350.0  60 months     12.69   

   installment grade sub_grade            emp_title  ... disbursement_method  \
0        84.33     C        C5                  NaN  ...                Cash   
1       339.31     C        C1  AIR RESOURCES BOARD  ...                Cash   
2       109.43     E        E1      MKC Accounting   ...                Cash   
3       152.39     F        F2                  NaN  ...                Cash   
4       121.45     B        B5            Starbucks  ...                Cash   

  debt_settlement_flag  debt_settlement_flag_date issue_year repaid  \
0                    N                        NaN       2011      1   
1                    N                        NaN       2011      1   
2                    N                        NaN       2011      1   
3                    N                        NaN       2011      0   
4                    N                        NaN       2011      0   

  loan_term_year purpose_code home_ownership_code grade_code addr_state_code  
0              3           11                   4          2               2  
1              3            9                   4          2               0  
2              3            0                   4          4               0  
3              5           11                   3          5               0  
4              5            9                   4          1               5  

[5 rows x 66 columns]

In [ ]:
X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

# Variable objetivo o variable a predecir
y = prestamos_df["repaid"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size= 0.3, stratify=y )

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((13935, 7), (5973, 7), (13935,), (5973,))

In [ ]:
# Create Regressor with default properties
rfc1 = RandomForestClassifier(random_state=23)

# Fit estimator and display score
rfc1 = rfc1.fit(X_train, y_train)

# Precisión del modelo en la fase de entrenamiento
print("Precision del clasificador en fase de entrenamiento", rfc1.score(X_train, y_train) )


Precision del clasificador en fase de entrenamiento 1.0


In [ ]:
# Realizar una prediccion con los datos de prueba
y_pred = rfc1.predict(X_test)

# Crear un informe de texto que muestre las principales métricas de clasificación.
print("\nReporte del clasificador Random Forest sin balanceo de clases : \n",
      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))

print(f'\nMatriz de Confusion Random Forest sin balanceo de clases:\n', confusion_matrix(y_test, y_pred ))


Reporte del clasificador Random Forest sin balanceo de clases : 
               precision    recall  f1-score   support

   No Pagado       0.31      0.03      0.06       883
      Pagado       0.85      0.99      0.92      5090

    accuracy                           0.85      5973
   macro avg       0.58      0.51      0.49      5973
weighted avg       0.77      0.85      0.79      5973


Matriz de Confusion Random Forest sin balanceo de clases:
 [[  28  855]
 [  61 5029]]


- Se obtuvo un 85% de precisión sin balanceo de clases
- Solo 28 impagos reales de 883

In [ ]:

rfc2 = RandomForestClassifier(class_weight='balanced', random_state=23)
rfc2 = rfc2.fit(X_train, y_train)
y_pred = rfc2.predict(X_test)

print("Reporte del clasificador con balanceo de clases: \n", classification_report(y_test, y_pred,
                                                            target_names=["No Pagado", "Pagado"] ))
print('Matriz de Confusion con balanceo de clases \n' , confusion_matrix(y_test, y_pred))

Reporte del clasificador con balanceo de clases: 
               precision    recall  f1-score   support

   No Pagado       0.31      0.02      0.04       883
      Pagado       0.85      0.99      0.92      5090

    accuracy                           0.85      5973
   macro avg       0.58      0.51      0.48      5973
weighted avg       0.77      0.85      0.79      5973

Matriz de Confusion con balanceo de clases 
 [[  21  862]
 [  46 5044]]


Con el balanceo se obtuvo el mismo porcentaje y solo disminuyeron 7 impagos reales

In [ ]:
rfc3 = RandomForestClassifier(n_estimators=200, max_depth=7, class_weight='balanced', random_state=23)

# Fit estimator and display score
rfc3 = rfc3.fit(X_train, y_train)

y_pred = rfc3.predict(X_test)

print("Reporte del clasificador con balanceo de clases n_estimators=200, max_depth=7: \n",
                   classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))
print('Matriz de Confusion con balanceo de clases n_estimators=200, max_depth=7\n' , confusion_matrix(y_test, y_pred))

Reporte del clasificador con balanceo de clases n_estimators=200, max_depth=7: 
               precision    recall  f1-score   support

   No Pagado       0.22      0.59      0.32       883
      Pagado       0.90      0.64      0.75      5090

    accuracy                           0.63      5973
   macro avg       0.56      0.61      0.53      5973
weighted avg       0.80      0.63      0.68      5973

Matriz de Confusion con balanceo de clases n_estimators=200, max_depth=7
 [[ 521  362]
 [1834 3256]]


Se modificaron los parámetros del modelo
- Disminuyó el porcentaje a 63%
- 521 impagos reales de 883

In [ ]:

rfc4 = RandomForestClassifier(n_estimators=200, max_depth=10, class_weight='balanced', random_state=23)

# Fit estimator and display score
rfc4 = rfc4.fit(X_train, y_train)

y_pred = rfc4.predict(X_test)

print("Reporte del clasificador con balanceo de clases n_estimators=200, max_depth=10: \n",
                      classification_report(y_test, y_pred, target_names=["No Pagado", "Pagado"]))
print('Matriz de Confusion con balanceo de clases n_estimators=200, max_depth=10\n' , confusion_matrix(y_test, y_pred))

Reporte del clasificador con balanceo de clases n_estimators=200, max_depth=10: 
               precision    recall  f1-score   support

   No Pagado       0.23      0.42      0.30       883
      Pagado       0.88      0.75      0.81      5090

    accuracy                           0.71      5973
   macro avg       0.56      0.59      0.56      5973
weighted avg       0.79      0.71      0.74      5973

Matriz de Confusion con balanceo de clases n_estimators=200, max_depth=10
 [[ 373  510]
 [1249 3841]]


Se aumenta la profundidad a 10
- Aumento la precisión a 71%
- Solo 373 impagos reales de 883

In [ ]:
# ==========================================
# Gráfica de radar - Random Forest (Plotly)
# ==========================================
import plotly.graph_objects as go

# --- Nombres de los modelos ---
modelos = [
    "RF sin balanceo",
    "RF balanceado",
    "RF bal. n=200, depth=7",
    "RF bal. n=200, depth=10"
]

# --- Métricas reales ---
accuracy = [0.85, 0.85, 0.63, 0.71]
recall_no_pagado = [0.03, 0.01, 0.58, 0.40]
recall_pagado = [0.99, 1.00, 0.66, 0.78]
f1_macro = [0.49, 0.47, 0.54, 0.57]

# --- Ejes ---
metricas = ["Accuracy", "Recall No Pagado", "Recall Pagado", "F1 Macro"]

# --- Crear la figura ---
fig = go.Figure()

# Agregar cada modelo al radar
for i, modelo in enumerate(modelos):
    fig.add_trace(go.Scatterpolar(
        r=[accuracy[i], recall_no_pagado[i], recall_pagado[i], f1_macro[i], accuracy[i]],
        theta=metricas + [metricas[0]],
        fill='toself',
        name=modelo
    ))

# --- Personalización ---
fig.update_layout(
    title="Comparación de modelos Random Forest (Radar Plot)",
    polar=dict(
        radialaxis=dict(visible=True, range=[0, 1], tickvals=[0.2, 0.4, 0.6, 0.8, 1.0]),
    ),
    showlegend=True
)
fig.update_layout(width=800, height=600)
fig.show()

In [ ]:
# Búsqueda de hiperparámetros con GridSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

# Iniciamos con un Modelo base con balanceo de clases (para compensar desbalance)
rf = RandomForestClassifier(class_weight='balanced', random_state=42)

# --- Definimos el espacio de búsqueda ---
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [5, 7, 10 ],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2']
}

# --- Configuración de GridSearchCV ---
grid_search = GridSearchCV(
    estimator=rf,
    param_grid=param_grid,
    scoring='f1_macro',   # Se puede cambiar a 'recall_macro', 'accuracy', etc.
    cv=3,                 # 3 particiones para validación cruzada
    n_jobs=-1,            # usa todos los núcleos disponibles
    verbose=2             # para ver el progreso
)

# --- Ejecutar búsqueda ---
grid_search.fit(X_train, y_train)

# --- Mostrar los mejores parámetros ---
print("Mejores hiperparámetros encontrados:")
print(grid_search.best_params_)

print("\n Mejor puntaje promedio (validación cruzada):")
print(grid_search.best_score_)

# --- Entrenar modelo final con los mejores parámetros ---
best_rf = grid_search.best_estimator_
y_pred_best = best_rf.predict(X_test)

# --- Evaluar en conjunto de prueba ---
print("\n Reporte de Clasificación (modelo óptimo):")
print(classification_report(y_test, y_pred_best, target_names=["No Pagado", "Pagado"]))

print("Matriz de Confusión:")
print(confusion_matrix(y_test, y_pred_best))

Fitting 3 folds for each of 162 candidates, totalling 486 fits


KeyboardInterrupt: 

- El random forest combina las predicciones de varios modelos simples como el de árboles de decisión, utilizando los árboles que necesites.
- Ventajas: Alta precisión, resistencia al sobreajuste, manejo de datos complejos

- Limitaciones: Consumo de memoria, lento de entrenar

- El modelo es malo detectando los préstamos no pagados

- Diferencia de clases muy grande, son más los pagados que los no pagados por esa razón no es capaz de detectar de mejor manera los préstamos no pagados.

- Tener la misma cantidad de clases

## Suport Vector Machine (SVM)

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go

# Importamos librerias necesarias para Random Forest Classifier
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, precision_recall_curve

In [ ]:

prestamos_df = pd.read_csv('prestamos_ok.csv')
prestamos_df.head()

Unnamed: 0  loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
0           0       2400         2400           2400.0  36 months     15.96   
1           1      10000        10000          10000.0  36 months     13.49   
2           2       3000         3000           3000.0  36 months     18.64   
3           3       5600         5600           5600.0  60 months     21.28   
4           4       5375         5375           5350.0  60 months     12.69   

   installment grade sub_grade            emp_title  ... disbursement_method  \
0        84.33     C        C5                  NaN  ...                Cash   
1       339.31     C        C1  AIR RESOURCES BOARD  ...                Cash   
2       109.43     E        E1      MKC Accounting   ...                Cash   
3       152.39     F        F2                  NaN  ...                Cash   
4       121.45     B        B5            Starbucks  ...                Cash   

  debt_settlement_flag  debt_settlement_flag_date issue_year repaid  \
0                    N                        NaN       2011      1   
1                    N                        NaN       2011      1   
2                    N                        NaN       2011      1   
3                    N                        NaN       2011      0   
4                    N                        NaN       2011      0   

  loan_term_year purpose_code home_ownership_code grade_code addr_state_code  
0              3           11                   4          2               2  
1              3            9                   4          2               0  
2              3            0                   4          4               0  
3              5           11                   3          5               0  
4              5            9                   4          1               5  

[5 rows x 66 columns]

In [ ]:
X = prestamos_df[['int_rate', 'loan_amnt','funded_amnt', 'annual_inc', 'dti', 'grade_code',
                 'revol_util']]

# Variable objetivo o variable a predecir
y = prestamos_df["repaid"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size= 0.3, stratify=y )


In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((13935, 7), (5973, 7), (13935,), (5973,))

In [ ]:
# --- Normalización de variables ---
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# --- Inicializar y entrenar el modelo SVM con balanceo de clases y probabilidad ---
svm_model = SVC(random_state=42, class_weight='balanced', probability=True)
svm_model.fit(X_train_scaled, y_train)

SVC(class_weight='balanced', probability=True, random_state=42)

Se hizo la normalizacion de las variables con StandardScaler

In [ ]:
# --- Predicción ---
y_pred_svm = svm_model.predict(X_test_scaled)

print("Reporte de Clasificación (SVM):")
print(classification_report(y_test, y_pred_svm, target_names=["No Pagado", "Pagado"]))

print("Matriz de Confusión (SVM):")
print(confusion_matrix(y_test, y_pred_svm))

Reporte de Clasificación (SVM):
              precision    recall  f1-score   support

   No Pagado       0.22      0.67      0.33       883
      Pagado       0.91      0.57      0.70      5090

    accuracy                           0.59      5973
   macro avg       0.56      0.62      0.52      5973
weighted avg       0.81      0.59      0.65      5973

Matriz de Confusión (SVM):
[[ 596  287]
 [2164 2926]]


- Se obtuvo un 59% de precisión
- 596 de impagos reales de 883

In [ ]:
##  Probabilidades de la clase positiva ##

from sklearn.metrics import roc_curve, roc_auc_score
import plotly.graph_objects as go

# --- Probabilidades de la clase positiva ("Pagado" = 1) ---
y_prob_svm = svm_model.predict_proba(X_test_scaled)[:, 1]

# --- Calcular AUC-ROC ---
auc = roc_auc_score(y_test, y_prob_svm)
print(f"\n AUC-ROC: {auc:.3f}")

# --- Curva ROC ---
fpr, tpr, thresholds = roc_curve(y_test, y_prob_svm)

# --- Gráfica interactiva con Plotly ---
fig = go.Figure()
fig.add_trace(go.Scatter(x=fpr, y=tpr, mode='lines', name='SVM', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=[0, 1], y=[0, 1], mode='lines', name='Aleatorio', line=dict(dash='dash')))
fig.update_layout(
    title=f'Probabilidades de la clase positiva - Curva ROC (AUC = {auc:.3f})',
    xaxis_title='Tasa de Falsos Positivos (1 - Especificidad)',
    yaxis_title='Tasa de Verdaderos Positivos (Sensibilidad)',
    width=700, height=500
)
fig.show()



 AUC-ROC: 0.667


In [ ]:

# --- Probabilidades de la clase negativa ("Pagado" = 0) ---
y_prob_svm = svm_model.predict_proba(X_test_scaled)[:, 0]

# --- Calcular AUC-ROC ---
auc = roc_auc_score(y_test, y_prob_svm)
print(f"\nAUC-ROC: {auc:.3f}")

# --- Curva ROC ---
fpr, tpr, thresholds = roc_curve(y_test, y_prob_svm)

# --- Gráfica interactiva con Plotly ---
fig = go.Figure()
fig.add_trace(go.Scatter(x=fpr, y=tpr, mode='lines', name='SVM', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=[0, 1], y=[0, 1], mode='lines', name='Aleatorio', line=dict(dash='dash')))
fig.update_layout(
    title=f'Probabilidades de la clase negativa - Curva ROC (AUC = {auc:.3f})',
    xaxis_title='Tasa de Falsos Positivos (1 - Especificidad)',
    yaxis_title='Tasa de Verdaderos Positivos (Sensibilidad)',
    width=700, height=500
)
fig.show()


AUC-ROC: 0.333


In [ ]:
# ===============================================
# 📈 Curvas ROC para ambas clases (positiva y negativa)
# ===============================================

from sklearn.metrics import roc_curve, roc_auc_score
import plotly.graph_objects as go

# --- Probabilidades predichas por el modelo ---
# Asegúrate de tener esto antes:
# svm_model.fit(X_train_scaled, y_train)
# y_prob_svm = svm_model.predict_proba(X_test_scaled)

# Calcular probabilidades y AUC para ambas clases
y_prob = svm_model.predict_proba(X_test_scaled)

# Clase positiva (Pagado = 1)
fpr_pos, tpr_pos, _ = roc_curve(y_test, y_prob[:, 1], pos_label=1)
auc_pos = roc_auc_score(y_test, y_prob[:, 1])

# Clase negativa (No Pagado = 0)
fpr_neg, tpr_neg, _ = roc_curve(y_test, y_prob[:, 0], pos_label=0)
auc_neg = roc_auc_score(y_test, y_prob[:, 0])

# --- Graficar con Plotly ---
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=fpr_pos, y=tpr_pos,
    mode='lines',
    name=f'Clase positiva (Pagado) - AUC = {auc_pos:.3f}',
    line=dict(color='blue', width=2)
))

fig.add_trace(go.Scatter(
    x=fpr_neg, y=tpr_neg,
    mode='lines',
    name=f'Clase negativa (No Pagado) - AUC = {auc_neg:.3f}',
    line=dict(color='green', width=2)
))

# Línea diagonal (modelo aleatorio)
fig.add_trace(go.Scatter(
    x=[0, 1], y=[0, 1],
    mode='lines',
    name='Aleatorio',
    line=dict(color='red', width=1, dash='dash')
))

# --- Personalización del gráfico ---
fig.update_layout(
    title="Comparación de Curvas ROC - Clases Positiva y Negativa",
    xaxis_title="Tasa de Falsos Positivos (1 - Especificidad)",
    yaxis_title="Tasa de Verdaderos Positivos (Sensibilidad)",
    width=900, height=600,
    legend=dict(x=0.6, y=0.05)
)

fig.show()


In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer, fbeta_score, classification_report, confusion_matrix
import numpy as np

# Definimos el F2-score como métrica personalizada (recall más importante que precisión)
f2_scorer = make_scorer(fbeta_score, beta=2, pos_label=0)

# Modelo base
svm = SVC(probability=False, class_weight='balanced', random_state=42)

# Definimos los hiperparámetros a explorar
param_grid = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf', 'poly'],
    'gamma': ['scale', 'auto'],
    'degree': [2, 3],  # solo afecta al kernel 'poly'
}

# Configuración del GridSearchCV
grid_search = GridSearchCV(
    estimator=svm,
    param_grid=param_grid,
    scoring=f2_scorer,  # optimizamos el F2-score
    cv=5,               # validación cruzada de 5 particiones
    verbose=2,
    n_jobs=-1           # usa todos los núcleos disponibles
)

# Entrenamiento
print("Entrenando GridSearchCV para SVM...\n")
grid_search.fit(X_train_scaled, y_train)

# Resultados
print("\n Búsqueda finalizada.")
print("Mejores hiperparámetros encontrados:")
print(grid_search.best_params_)

print("\nMejor puntaje promedio (F2-score):")
print(round(grid_search.best_score_, 4))

# Evaluación con los mejores parámetros
svm_optimo = grid_search.best_estimator_
y_pred_opt = svm_optimo.predict(X_test_scaled)

print("\n Reporte de Clasificación (SVM optimizado para impagos):\n")
print(classification_report(y_test, y_pred_opt, target_names=["No Pagado", "Pagado"]))

print("Matriz de Confusión:")
print(confusion_matrix(y_test, y_pred_opt))

Entrenando GridSearchCV para SVM...

Fitting 5 folds for each of 36 candidates, totalling 180 fits

 Búsqueda finalizada.
Mejores hiperparámetros encontrados:
{'C': 0.1, 'degree': 2, 'gamma': 'scale', 'kernel': 'rbf'}

Mejor puntaje promedio (F2-score):
0.4734

 Reporte de Clasificación (SVM optimizado para impagos):

              precision    recall  f1-score   support

   No Pagado       0.21      0.67      0.32       883
      Pagado       0.91      0.57      0.70      5090

    accuracy                           0.59      5973
   macro avg       0.56      0.62      0.51      5973
weighted avg       0.81      0.59      0.65      5973

Matriz de Confusión:
[[ 592  291]
 [2179 2911]]


Se buscaron los mejores parámetros para el modelo
- Se obtuvo un 59% de precisión
- 592 de impagos reales de 883

- La idea fundamental detrás de SVM para clasificación es encontrar el "mejor" hiperplano que separe las clases de datos en un espacio multidimensional. Este hiperplano se elige de manera que el margen (la distancia entre el hiperplano y los puntos de datos más cercanos de cada clase, conocidos como vectores de soporte) sea lo más grande posible.

- Ventajas: Versátil con Kernels, resistente al sobreajuste

- Limitaciones: es sensible a la elección del kernel, No Escala Bien con Grandes Conjuntos de Datos.

- No detecta bien los casos en donde el préstamo no fue pagado

- Desbalance de clases, selección de variables insuficientes, sensibilidad

- Mejorar la igualdad de clases en los datos, selección de variables más eficaces.

##Gráfico radar o spider chart

In [5]:
model_names = ['KNN (SMOTE)', 'Regresión Logística (escalado)', 'Árbol de Decisión (optimizado)', 'Random Forest (sintonizado)', 'SVM (optimizado)']

# Extraer las métricas finales de cada modelo

# KNN (SMOTE) - from cell bd280cd8
knn_accuracy = 0.69
knn_recall_no_pagado = 0.24
knn_recall_pagado = 0.78
knn_f1_macro = 0.50

# Regresión Logística (escalado) - from cells whY9jZIQogoQ and I5F88G2nokSE
rl_accuracy = 0.61
rl_recall_no_pagado = 0.61
rl_recall_pagado = 0.62
rl_f1_macro = 0.53

# Árbol de Decisión (optimizado) - from cell uV8q3ax8vNAe
dt_accuracy = 0.54
dt_recall_no_pagado = 0.73
dt_recall_pagado = 0.51
dt_f1_macro = 0.49

# Random Forest (sintonizado) - from cell 179SniMn6Xq5 (rfc4 with n_estimators=200, max_depth=10, class_weight='balanced')
rf_accuracy = 0.71
rf_recall_no_pagado = 0.42
rf_recall_pagado = 0.75
rf_f1_macro = 0.56

# SVM (optimizado) - from cell L7NiGqVe87F7
svm_accuracy = 0.59
svm_recall_no_pagado = 0.67
svm_recall_pagado = 0.57
svm_f1_macro = 0.51

# Almacenar las métricas en las listas especificadas
accuracy_scores = [
    knn_accuracy,
    rl_accuracy,
    dt_accuracy,
    rf_accuracy,
    svm_accuracy
]

recall_no_pagado_scores = [
    knn_recall_no_pagado,
    rl_recall_no_pagado,
    dt_recall_no_pagado,
    rf_recall_no_pagado,
    svm_recall_no_pagado
]

recall_pagado_scores = [
    knn_recall_pagado,
    rl_recall_pagado,
    dt_recall_pagado,
    rf_recall_pagado,
    svm_recall_pagado
]

f1_macro_scores = [
    knn_f1_macro,
    rl_f1_macro,
    dt_f1_macro,
    rf_f1_macro,
    svm_f1_macro
]

print("Métricas extraídas:")
print(f"Modelos: {model_names}")
print(f"Accuracy: {accuracy_scores}")
print(f"Recall 'No Pagado': {recall_no_pagado_scores}")
print(f"Recall 'Pagado': {recall_pagado_scores}")
print(f"F1-Macro: {f1_macro_scores}")

Métricas extraídas:
Modelos: ['KNN (SMOTE)', 'Regresión Logística (escalado)', 'Árbol de Decisión (optimizado)', 'Random Forest (sintonizado)', 'SVM (optimizado)']
Accuracy: [0.69, 0.61, 0.54, 0.71, 0.59]
Recall 'No Pagado': [0.24, 0.61, 0.73, 0.42, 0.67]
Recall 'Pagado': [0.78, 0.62, 0.51, 0.75, 0.57]
F1-Macro: [0.5, 0.53, 0.49, 0.56, 0.51]


In [4]:
import plotly.graph_objects as go

# Ejes del radar (métricas)
metricas = ["Accuracy", "Recall No Pagado", "Recall Pagado", "F1 Macro"]

# Crear la figura del radar
fig = go.Figure()

# Añadir cada modelo al radar
for i, model_name in enumerate(model_names):
    fig.add_trace(go.Scatterpolar(
        r=[
            accuracy_scores[i],
            recall_no_pagado_scores[i],
            recall_pagado_scores[i],
            f1_macro_scores[i],
            accuracy_scores[i] # Cerrar el polígono
        ],
        theta=metricas + [metricas[0]], # Cerrar el polígono
        fill='toself',
        name=model_name
    ))

# Personalización del gráfico
fig.update_layout(
    title="Comparación de Modelos ML (Radar Plot)",
    polar=dict(
        radialaxis=dict(
            visible=True,
            range=[0, 1], # Las métricas suelen estar entre 0 y 1
            tickvals=[0.2, 0.4, 0.6, 0.8, 1.0]
        )
    ),
    showlegend=True,
    width=800,
    height=600
)

fig.show()

In [3]:
import pandas as pd

# Recopilar las métricas específicas
model_names = [
    'KNN (SMOTE)',
    'Regresión Logística (escalado)',
    'Árbol de Decisión (optimizado)',
    'Random Forest (sintonizado)',
    'SVM (optimizado)'
]

# Accuracy
accuracy = [0.69, 0.61, 0.54, 0.71, 0.59]

# Recall para 'No Pagado' (Clase 0)
recall_no_pagado = [0.24, 0.61, 0.73, 0.42, 0.67]

# F1-score para 'No Pagado' (Clase 0) - extraídos de los reportes de clasificación correspondientes
f1_no_pagado = [0.20, 0.32, 0.32, 0.30, 0.32]

# AUC - solo disponible para SVM en el cuaderno actual, otros N/A
auc_scores = ['N/A', 'N/A', 'N/A', 'N/A', 0.667] # AUC para SVM (clase 'Pagado' = 1)

# Crear un DataFrame para la tabla comparativa
comparative_df = pd.DataFrame({
    'Modelo': model_names,
    'Accuracy': accuracy,
    'Recall (No Pagado)': recall_no_pagado,
    'F1-score (No Pagado)': f1_no_pagado,
    'AUC': auc_scores
})

print("\nTabla Comparativa de Modelos:\n")
display(comparative_df)


Tabla Comparativa de Modelos:



Modelo  Accuracy  Recall (No Pagado)  \
0                     KNN (SMOTE)      0.69                0.24   
1  Regresión Logística (escalado)      0.61                0.61   
2  Árbol de Decisión (optimizado)      0.54                0.73   
3     Random Forest (sintonizado)      0.71                0.42   
4                SVM (optimizado)      0.59                0.67   

   F1-score (No Pagado)    AUC  
0                  0.20    N/A  
1                  0.32    N/A  
2                  0.32    N/A  
3                  0.30    N/A  
4                  0.32  0.667

Preguntas:

- ¿Qué modelo resultó más útil y por qué?

Árbol de Decisión (optimizado): Con un Recall (No Pagado) de 0.73, este modelo es el mejor identificando a los clientes que no pagarán sus préstamos. Esto es extremadamente valioso para minimizar las pérdidas financieras, ya que permite al banco actuar proactivamente (por ejemplo, rechazando el préstamo o ajustando las condiciones).
SVM (optimizado): Le sigue de cerca con un Recall (No Pagado) de 0.67. También es muy eficaz detectando impagos.
- ¿Qué variables parecen tener más peso en la predicción?

int_rate (tasa de interés): A menudo es un reflejo directo del riesgo inicial percibido del prestatario; tasas más altas suelen ir asociadas a un mayor riesgo.grade_code (código de calificación crediticia): Es una evaluación consolidada del perfil de riesgo del prestatario, y generalmente es un predictor muy fuerte.annual_inc (ingreso anual)

- ¿Qué decisiones se pueden tomar con esta información en un banco o fintech?  

Decisiones de Aprobación de Crédito, Establecimiento de Tasas de Interés y Condiciones,Límites de Crédito.